design/002-secret-generator-group.md

```diff
@@ -0,0 +1,277 @@
+
+#### Example CRD
+
+The following custom resource `Kind=VaultPKIBackend` contains all the parameters for issuing a certificate via HashiCorp Vault PKI engine.
+
+```yaml
+apiVersion: generator.external-secrets.io/v1alpha1
+kind: VaultPKIBackend
+metadata:
+  name: my-vault-pki-creds
+spec:
+  # specific params for generating
+  commonName: example.com
+  altNames:
+  - example.org
+  - localhost
+  # ... more opts
+  auth: {} # service account etc. etc.
+```
+
+This resource can be **embedded or referenced** from an ExternalSecret or PushSecret.
+The ExternalSecret controller generates the secret value and it can be post-processed with
+existing features like template or key-rewrite.
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  # ...
+  dataFrom:
+
+    # embedded
+    - generator:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: VaultPKIBackend
+        spec:
+          # same as above
+          commonName: example.com
+          altNames:
+          - example.org
+          - localhost
+          # ... more opts
+          auth: {} # service account etc. etc.issuerRef: "default"
+
+    # the generator also can be referenced like here
+    - generatorRef:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: VaultPKIBackend
+        name: foobar
+```
+
+
+<details>
+<summary>PushSecret example</summary>
+
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: PushSecret
+metadata:
+  name: "hello-world"
+spec:
+  secretStoreRefs: [] # ...
+  selector:
+    # we can consider supporting multiple generators per PushSecret
+    # but need to take care of key collision
+    generatorRef:
+      apiVersion: generator.external-secrets.io/v1alpha1
+      kind: VaultPKIBackend
+      name: foobar
+  data: {} # ...
+```
+</details>
+
+<details>
+<summary>Password Generator Example</summary>
+
+```yaml
+apiVersion: generator.external-secrets.io/v1alpha1
+kind: Password
+metadata:
+  name: foo
+spec:
+  characterSet: "..." # TBD
+  length: 42
+---
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  # ...
+  dataFrom:
+    - generatorRef:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: Password
+        name: foo
+```
+</details>
+
+<details>
+<summary>Example: multiple generators with key rewrite</summary>
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  # ...
+  dataFrom:
+    # using multiple generators
+    # this will produce the following values:
+    # foo-username
+    # foo-password
+    # bar-username
+    # bar-password
+    - generatorRef:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: ECR
+        name: foo
+      rewrite:
+      - regexp:
+         source: .+
+         target: foo-$1
```

Comment:
Would the output of the generatorRef be always a `map[string]string`?

Reply:
`spec.data` wouldn't really work well with a generator that outputs a `map[string]string` as we would have to serialize that map into a string. That feels odd, especially with the `rewrite` feature in place it shouldn't be needed.

Reply:
yes, I guess the appropriate type anyways would be a `map[string][]byte` right :stuck_out_tongue: ? I was just thinking that this `sourceRef` thing could add some general value to all operations we have. I agree it can be tricky to do without adding yet another GetSecret/GetSecretMap combo when there is a generator in place.